test/e2e/expirable-airdrop.spec.ts

```diff
@@ -0,0 +1,250 @@
+    return getLeaf(address, toClaim);
+  });
+
+  const tree = new MerkleTree(leaves, ethers.utils.keccak256, { sortPairs: true });
+  const root = tree.getHexRoot();
+
+  return [root, tree];
+}
+
+describe('Expirable airdrop', () => {
+  let nowTimestamp: number;
+  const oneMonth: number = 30 * 24 * 60 * 60; // one month in seconds
+  let expirationTimestamp: number;
+  let accounts: any[];
+  let deployer: any;
+  let root: any;
+  let tree: any;
+  let token: any;
+  let expirableAirdrop: any;
+  let airdropAmount: number = 10;
+  let airdropAmountBN: BigNumber = toUnit(airdropAmount);
+  let snapshot: any;
+  let alice: SignerWithAddress;
+  let bob: SignerWithAddress;
+  let carl: SignerWithAddress;
+  let random: SignerWithAddress;
+  const randomAddress: string = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045';
+
+  async function prepareAirdrop() {
+    let toDeposit = toUnit(airdropAmount * (accounts.length + 1));
+    await token.mint(expirableAirdrop.address, toDeposit);
+  }
+
+  async function claim(sender: SignerWithAddress, claimee: SignerWithAddress) {
+    let leaf = getLeaf(claimee.address, airdropAmountBN);
+    let proof = tree.getHexProof(leaf);
+
+    await expirableAirdrop.connect(sender).claimAndSendToClaimee(claimee.address, airdropAmountBN, proof);
+  }
+
+  before(async () => {
+    nowTimestamp = await time.latest();
+    expirationTimestamp = nowTimestamp + 3 * oneMonth; // 3 months
+    [deployer, ...accounts] = await ethers.getSigners();
+
+    [root, tree] = createMerkleRoot(accounts, airdropAmountBN);
+
+    // deploy erc20 mock
+    const TokenFactory = await ethers.getContractFactory('ERC20Mock');
+    token = await TokenFactory.deploy('Token', 'TKN');
+
+    // deploy contract
+    const ExpirableAirdropFactory = await ethers.getContractFactory('ExpirableAirdrop');
+    expirableAirdrop = await ExpirableAirdropFactory.deploy(deployer.address, token.address, expirationTimestamp, root);
+
+    // load accounts
+    alice = accounts[1];
+    bob = accounts[2];
+    carl = accounts[3];
+    await impersonateAccount(randomAddress);
+    await setBalance(randomAddress, toUnit(10000));
+    random = await ethers.getSigner(randomAddress);
+
+    // save snapshot for later
+    snapshot = await takeSnapshot();
+  });
+
+  beforeEach(async () => {
+    await snapshot.restore();
+  });
+
+  when('deposit', () => {
+    let toDeposit: BigNumber;
+
+    given(async () => {
+      toDeposit = toUnit(airdropAmount * (accounts.length + 1));
+
+      // transfer some tokens
+      await token.mint(deployer.address, toDeposit);
+    });
+
+    then('airdrop is deposited', async () => {
+      // approve
+      await token.approve(expirableAirdrop.address, ethers.constants.MaxUint256);
+
+      // deposit
+      await expirableAirdrop.depositTokens(toDeposit);
+
+      // expect
+      expect(await token.balanceOf(expirableAirdrop.address)).equal(toDeposit);
+    });
+  });
+
+  when('claim and send to claimee', () => {
```

Comment:
I would suggest a slightly different approach. Something like this:
```
describe('claimAndSendToClaimee', () => {
    let leaf: any;
    let proof: any;
    given(async () => {
        await prepareAirdrop();
        leaf = getLeaf(alice.address, airdropAmountBN);
        proof = tree.getHexProof(leaf);
    });
    when('airdrop is claimed by claimee', () => {
        let tx: TransactionResponse
        given(() => {
            tx = await expirableAirdrop.connect(alice).claimAndSendToClaimee(alice.address, airdropAmountBN, proof);
        })
        then('tokens are transferred to claimee', () => {
            // Checks balance
        })
        then('event is emitted', () => {
            // Checks event in tx
        })
        then('claimee is marked as claimed', () => {
            // Checks that claimee is marked as claimed
        })
    }
    when('airdrop is claimed on behalf of claimee', () => {
        ...
    })
    when('airdrop is already claimed', () => {
        given(() => {
            // Claim once
        })
        then('claiming again will result in a revert', () => {
            ...
        })
    })
})
```
I believe this will make tests clearer and easier to read, since there is a clear separation of what is executed vs what the result is expected to be

In some cases it can lead to repeated code. For example, in this case, the assertions in the first two tests would be the same. In those cases, you can create a function like this:
```
function thenAirdropIsClaimed(claimee: string, reciever: string, amount: BigNumberish) {
    then('tokens are transferred to claimee', () => {
        // Checks balance
    })
    then('event is emitted', () => {
        // Checks event in tx
    })
    then('claimee is marked as claimed', () => {
        // Checks that claimee is marked as claimed
    })
}
```
And you can simply call it from inside those two tests. 

I know it can be a lot, and I know that I'm very special about tests, but I've learned that it's very important that they are clear and readable. By doing so, it should take 5 seconds to map assertions to state modifications, and you will be able to tell if you are testing everything you should be testing